src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/BackgroundWriter.java

```diff
@@ -0,0 +1,194 @@
+     * See "Java Concurrency In Practice", by Brian Goetz et al,
+     * 8.3.3: Saturation Policies (last paragraph) and
+     * Listing 8.4: Using a Semaphore to throttle task submission.
+     */
+    private final Semaphore semaphore;
+
+    /**
+     * The time at which the writer is initialized.
+     */
+    private long start;
+
+    /**
+     * The cumulative time spent writing documents
+     */
+    private double taskBusy = 0;
+
+    /**
+     * Whether to report additional information, such as the overall utilization of the
+     * background thread.
+     */
+    private boolean verbose;
+
+    /**
+     * Creates a {@code BackgroundWriter}.
+     *
+     * @implNote
+     * The writer uses a {@link Executors#newFixedThreadPool fixed thread pool} of
+     * {@value #BACKGROUND_THREADS} background threads, and a blocking queue of up to
+     * {@value #QUEUED_TASKS} queued tasks.
+     *
+     * @param messages used to write out any error messages and other output
+     * @param verbose  if true, writes out debugging info about the utilization of the background thread
+     */
+    public BackgroundWriter(Messages messages, boolean verbose) {
+        this.messages = messages;
+        this.verbose = verbose;
+        executor = Executors.newFixedThreadPool(BACKGROUND_THREADS);
+        semaphore = new Semaphore(QUEUED_TASKS + BACKGROUND_THREADS);
+        start = System.currentTimeMillis();
+    }
+
+    /**
+     * Writes the given document at some point in the future, using the internal executor.
+     * A semaphore is used to throttle the number of requests, although in practice this
+     * rarely takes effect, because it is quicker to write documents than to generate them.
+     *
+     * @param doc  the document to be written
+     * @param file the file to which to write the document
+     */
+    public void writeLater(HtmlDocument doc, DocFile file) {
+        try {
+            semaphore.acquire();
+            try {
+                executor.execute(() -> {
+                    try {
+                        long taskStart  = System.currentTimeMillis();
+                        doc.write(file);
+                        long taskEnd  = System.currentTimeMillis();
+                        taskBusy += (taskEnd - taskStart);
+                    } catch (DocFileIOException e) {
+                        error(e);
+                    } finally {
+                        semaphore.release();
+                    }
+                });
+            } catch (RejectedExecutionException e) {
+                semaphore.release();
+            }
+        } catch (InterruptedException e) {
+            messages.error("doclet.exception.interrupted");
+        }
+
+    }
+
+    /**
+     * Shuts down the internal executor service, releasing all resources.
+     */
+    public void finish() {
+        try {
+            executor.shutdown();
+            executor.awaitTermination(5, TimeUnit.MINUTES);
```

Comment:
I don't understand why you think it is important to check `taskBusy`. If nothing else, this is only in the verbose/utilization code.

Reply:
`taskBusy` is accessed from multiple threads. Since those accesses are conflicting they should be correctly synchronized, which currently they are not. Also see: https://github.com/openjdk/jdk/pull/2665#issuecomment-783378558

Reply:
Understood; I'm changing it `AtomicLong` and using `AtomicLong.addAndget`